roofit/multiprocess/test/test_Queue.cxx

```diff
@@ -0,0 +1,135 @@
+/*
+* Project: RooFit
+* Authors:
+*   PB, Patrick Bos, Netherlands eScience Center, p.bos@esciencecenter.nl
+*
+* Copyright (c) 2022, CERN
+*
+* Redistribution and use in source and binary forms,
+* with or without modification, are permitted according to the terms
+* listed in LICENSE (http://roofit.sourceforge.net/license.txt)
+*/
+
+#include "RooFit/MultiProcess/Config.h"
+
+#include <algorithm>  // std::equal
+
+#include "OrderTrackingJob.h"
+#include "gtest/gtest.h"
+
+TEST(FIFOQueue, TaskOrder)
+{
+   // one worker so we can easily check order, because of deterministic serial task execution on one worker
+   RooFit::MultiProcess::Config::setDefaultNWorkers(1);
+
+   std::size_t n_tasks = 10;
+   OrderTrackingJob job(n_tasks, 0);
+
+   std::vector<std::size_t> expected_order{0, 1, 2, 3, 4, 5, 6, 7, 8, 9};
+
+   job.do_the_job();
+
+   auto job_id = job.get_job_id();
+   EXPECT_TRUE(std::equal(job.received_task_order[job_id].cbegin(),
+                          job.received_task_order[job_id].cend(),
+                          expected_order.cbegin()));
+}
+
+TEST(PriorityQueue, TaskOrder)
+{
+   // one worker so we can easily check order, because of deterministic serial task execution on one worker
+   RooFit::MultiProcess::Config::setDefaultNWorkers(1);
+
+   EXPECT_TRUE(RooFit::MultiProcess::Config::Queue::setQueueType(RooFit::MultiProcess::Config::Queue::QueueType::Priority));
+
+   std::size_t n_tasks = 10;
+   OrderTrackingJob job(n_tasks, 10000);
+
+   std::vector<RooFit::MultiProcess::Task> reverse_order{9, 8, 7, 6, 5, 4, 3, 2, 1, 0};
+
+   RooFit::MultiProcess::Config::Queue::suggestTaskOrder(job.get_job_id(), reverse_order);
+
+   job.do_the_job();
+
+   auto job_id = job.get_job_id();
+
+   // We expect the first executed task to be task 0, sometimes task 1. This
+   // is because the worker will immediately steal a task once one is on the
+   // queue, so usually that will be task 0. Sometimes, the worker's request
+   // comes just before task 0 was added, so then the queue first rejects the
+   // work request, adds task 0, and then maybe task 1 is added in the meantime
+   // while the worker still has to send its next work request, so then it gets
+   // 1 (in rare occasions, it could get task 2 or an even later one; if this
+   // happens too often we should modify this test).
+   // The rest of the tasks, we expect to be in the suggested order, because we
+   // let the first task sleep sufficiently long for the other tasks to be sent
+   // to the queue before the next worker's request for work.
+   // The last task, we don't check, because it might be either 1 or 0,
+   // depending on which task was done first.
+   EXPECT_TRUE(std::equal(job.received_task_order[job_id].cbegin() + 1,
+                          job.received_task_order[job_id].cend() - 1,
+                          reverse_order.cbegin()));
```

Comment:
```suggestion
   // We expect the first executed task to be task 0, sometimes task 1. This is
   // because the worker will immediately steal a task once one is on the
   // queue, so usually that will be task 0. Sometimes, the worker's request
   // comes just before task 0 was added, so then the queue first rejects the
   // work request, adds task 0, and then maybe task 1 is added in the meantime
   // while the worker still has to send its next work request, so then it gets
   // 1 or an evan later task. So when comparing to the sugested order, we have
   // to move this stolen task to the front.
   auto const& recieved = job.received_task_order.at(job.get_job_id());
   auto stolen_task_iter = std::find(order.begin(), order.end(), recieved.front());
   std::rotate(order.begin(), stolen_task_iter, stolen_task_iter + 1);

   EXPECT_TRUE(std::equal(recieved.cbegin(), recieved.cend(), order.cbegin()));
```

I think we can solve the problem with the test by always rotating the stolen task to the front of the reverence order, no matter which one it was.

Not that I have renamed `reverse_order` to `order` for code brevity and because the order will not be strictly reverse anymore after the rotation.

Also, the same could be done for the `TaskPriority` test.

Also, as a sidenote: if you want to get an element from a map where you are sure it exists, use `at()` (`operator[]` will add a new default-initialized element if the key doesn't exists, and this has caused me headache many times already...)

Reply:
Brilliant! Thanks for this fix, it seems to be working perfectly (and indeed I cannot see why it wouldn't).